File: python/image_tools/download_assets.py
#!/usr/bin/env python3
"""
🌍 Download de Imagens e GIFs de Combate

Baixa imagens estáticas e GIFs de animação diretamente da internet
para os personagens e bosses do gacha-game. Também atualiza o banco
SQLite com as URLs locais dos assets.

Funcionalidades:
  - Baixa imagens estáticas de personagens (160x160)
  - Baixa GIFs de animação (idle, attack, defend, skill, hit, victory, defeat)
  - Baixa imagens de bosses
  - Atualiza o schema do banco (image_url, gif_*_url)
  - Cria índices de busca por anime/elemento

Uso:
  python download_assets.py                    # Baixa tudo
  python download_assets.py --character frieren # Só Frieren
  python download_assets.py --dry-run          # Mostra o que faria sem baixar
"""

import os
import sys
import json
import sqlite3
import time
import urllib.request
import urllib.error
from pathlib import Path

# --- Configuração ---
BASE_DIR = Path(__file__).resolve().parent.parent.parent
DB_PATH = BASE_DIR / "database.sqlite"
IMAGES_DIR = BASE_DIR / "assets" / "images"
GIFS_DIR = BASE_DIR / "assets" / "gifs"
SOUNDS_DIR = BASE_DIR / "assets" / "sounds"

# Serviço de imagens placeholder (placehold.co)
PLACEHOLDER_BASE = "https://placehold.co/160x160"

# Estados de animação de combate
COMBAT_STATES = ["idle", "attack", "defend", "skill", "hit", "victory", "defeat"]

# Banco de personagens (id → nome, anime, emoji)
CHARACTERS = {
    1: {"name": "Frieren", "anime": "Frieren", "emoji": "🧙‍♀️", "rarity": "secret",
        "element": "magia", "role": "attacker"},
    2: {"name": "Gilgamesh", "anime": "Gilgamesh", "emoji": "👑", "rarity": "secret",
        "element": "ouro", "role": "attacker"},
    3: {"name": "Madoka", "anime": "Madoka", "emoji": "🌸", "rarity": "secret",
        "element": "luz", "role": "support"},
    4: {"name": "Aizen", "anime": "Bleach", "emoji": "👁️", "rarity": "mythic",
        "element": "trevas", "role": "support"},
    5: {"name": "Ulquiorra", "anime": "Bleach", "emoji": "🌑", "rarity": "legendary",
        "element": "vazio", "role": "attacker"},
    6: {"name": "Kars", "anime": "JoJo", "emoji": "🦴", "rarity": "mythic",
        "element": "pedra", "role": "tank"},
    7: {"name": "DIO", "anime": "JoJo", "emoji": "👑", "rarity": "legendary",
        "element": "trevas", "role": "attacker"},
    8: {"name": "Renji", "anime": "Bleach", "emoji": "🔥", "rarity": "epic",
        "element": "fogo", "role": "attacker"},
    9: {"name": "Yusuke", "anime": "YuYu", "emoji": "👻", "rarity": "epic",
        "element": "espírito", "role": "attacker"},
    10: {"name": "Gon", "anime": "Hunter x Hunter", "emoji": "💚", "rarity": "rare",
         "element": "natureza", "role": "attacker"},
}

# Bosses
BOSSES = {
    1: {"name": "Aizen", "anime": "Bleach", "emoji": "👁️", "difficulty": "hard"},
    2: {"name": "Ulquiorra", "anime": "Bleach", "emoji": "🌑", "difficulty": "normal"},
    3: {"name": "Kars", "anime": "JoJo", "emoji": "🦴", "difficulty": "hard"},
    4: {"name": "DIO", "anime": "JoJo", "emoji": "👑", "difficulty": "nightmare"},
    5: {"name": "Kamina", "anime": "Tengen", "emoji": "🔥", "difficulty": "hard"},
    6: {"name": "Saitama", "anime": "One Punch", "emoji": "🥊", "difficulty": "nightmare"},
    7: {"name": "All For One", "anime": "MHA", "emoji": "🖤", "difficulty": "hard"},
    8: {"name": "Light Yagami", "anime": "Death Note", "emoji": "📓", "difficulty": "normal"},
    9: {"name": "Askeladd", "anime": "Vinland", "emoji": "⚔️", "difficulty": "hard"},
}

RARITY_COLORS_HEX = {
    "common": "95A5A6", "rare": "74B9FF", "epic": "4F9DDE",
    "legendary": "B388FF", "mythic": "FFC93C", "secret": "FF3B3B",
}


def get_db():
    """Abre conexão com o SQLite."""
    if not DB_PATH.exists():
        print(f"⚠️  Banco não encontrado em: {DB_PATH}")
        print("   Execute o seed do projeto primeiro.")
        return None
    conn = sqlite3.connect(str(DB_PATH))
    conn.row_factory = sqlite3.Row
    return conn


def download_image(url, filepath, retries=3, delay=0.5):
    """
    Baixa um arquivo da internet e salva localmente.
    Timeout de 10s, com retry.
    """
    filepath.parent.mkdir(parents=True, exist_ok=True)
    if filepath.exists() and filepath.stat().st_size > 100:
        return filepath  # já baixado

    for attempt in range(retries):
        try:
            req = urllib.request.Request(url, headers={"User-Agent": "GachaGame-AssetDownloader/1.0"})
            with urllib.request.urlopen(req, timeout=10) as resp:
                data = resp.read()
                if len(data) > 100:
                    filepath.write_bytes(data)
                    print(f"  📥 {filepath.relative_to(BASE_DIR)}")
                    return filepath
        except (urllib.error.URLError, Exception) as e:
            if attempt < retries - 1:
                time.sleep(delay)
            else:
                print(f"  ❌ Falhou: {url} → {e}")
    return None


def make_placeholder(name, emoji, rarity, bg_color=None, size=160):
    """
    Gera URL de placeholder via placehold.co.
    Para usar quando não conseguimos baixar da internet.
    """
    bg = bg_color or RARITY_COLORS_HEX.get(rarity, "888888")
    text = f"{emoji} {name}".replace(" ", "+").replace("👑", "").strip()
    text = f"{emoji} {name}"
    # placehold.co não suporta emojis bem, então usa o nome
    safe_name = name.replace(" ", "+")
    url = f"{PLACEHOLDER_BASE}/{size}x{size}/{bg}/FFFFFF?text={safe_name}"[:200]
    return url


def ensure_columns(conn):
    """
    Garante que as colunas de animação existam no schema.
    Adiciona image_url, gif_*_url às tabelas characters e bosses.
    """
    cols_to_check = {
        "characters": ["image_url", "gif_attack_url", "gif_defend_url",
                       "gif_skill_url", "gif_hit_url", "gif_victory_url", "gif_defeat_url"],
        "bosses": ["image_url", "gif_attack_url", "gif_defend_url"],
    }

    # Adiciona colunas se faltam
    for table, cols in cols_to_check.items():
        existing = {r[1] for r in conn.execute(f"PRAGMA table_info({table})").fetchall()}
        for col in cols:
            if col not in existing:
                conn.execute(f"ALTER TABLE {table} ADD COLUMN {col} TEXT")
                print(f"  ➕ Adicionada coluna {col} em {table}")
    conn.commit()


def update_character_urls(conn, char_id, char_data, asset_prefix="/assets/"):
    """Atualiza as URLs de imagem/GIF de um personagem no banco."""
    char_name = char_data["name"]
    rarity = char_data["rarity"]

    image_url = f"{asset_prefix}images/{char_name.lower()}.png"
    updates = {"image_url": image_url}

    for state in COMBAT_STATES:
        if state == "idle":
            updates["gif_attack_url"] = f"{asset_prefix}gifs/{char_name.lower()}_attack.gif"
        # O idle é a imagem estática; outros states são GIFs
        if state == "attack":
            updates["gif_attack_url"] = f"{asset_prefix}gifs/{char_name.lower()}_attack.gif"
        elif state == "defend":
            updates["gif_defend_url"] = f"{asset_prefix}gifs/{char_name.lower()}_defend.gif"
        elif state == "skill":
            updates["gif_skill_url"] = f"{asset_prefix}gifs/{char_name.lower()}_skill.gif"
        elif state == "hit":
            updates["gif_hit_url"] = f"{asset_prefix}gifs/{char_name.lower()}_hit.gif"
        elif state == "victory":
            updates["gif_victory_url"] = f"{asset_prefix}gifs/{char_name.lower()}_victory.gif"
        elif state == "defeat":
            updates["gif_defeat_url"] = f"{asset_prefix}gifs/{char_name.lower()}_defeat.gif"

    set_clauses = []
    values = []
    for col, val in updates.items():
        set_clauses.append(f"{col} = ?")
        values.append(val)
    values.append(char_id)

    conn.execute(
        f"UPDATE characters SET {', '.join(set_clauses)} WHERE id = ?",
        values,
    )
    conn.commit()
    print(f"  📝 Banco atualizado: {char_name} (id={char_id})")


def update_boss_urls(conn, boss_id, boss_data, asset_prefix="/assets/"):
    """Atualiza as URLs de imagem/GIF de um boss no banco."""
    boss_name = boss_data["name"]
    image_url = f"{asset_prefix}images/{boss_name.lower()}.png"
    attack_url = f"{asset_prefix}gifs/{boss_name.lower()}_attack.gif"
    defend_url = f"{asset_prefix}gifs/{boss_name.lower()}_defend.gif"

    conn.execute(
        "UPDATE bosses SET image_url = ?, gif_attack_url = ?, gif_defend_url = ? WHERE id = ?",
        (image_url, attack_url, defend_url, boss_id),
    )
    conn.commit()
    print(f"  📝 Banco atualizado: Boss {boss_name} (id={boss_id})")


def download_character_assets(char_id, char_data, dry_run=False):
    """Baixa imagem e GIFs de um personagem."""
    name = char_data["name"].lower()
    print(f"\n🔧 {char_data['name']} ({char_data['rarity']}) [{char_data['anime']}]")

    # 1. Imagem estática (idle)
    img_path = IMAGES_DIR / f"{name}.png"
    url = make_placeholder(char_data["name"], char_data["emoji"], char_data["rarity"])
    if not dry_run:
        download_image(url, img_path)
    else:
        print(f"  [DRY-RUN] Baixaria: {url} → {img_path}")

    # 2. GIFs de animação para cada estado
    for state in COMBAT_STATES:
        gif_path = GIFS_DIR / f"{name}_{state}.gif"
        # Usa placehold.co com texto indicando o estado
        state_label = {"idle": "Idle", "attack": "Attack", "defend": "Defend",
                       "skill": "Skill", "hit": "Hit", "victory": "Victory", "defeat": "Defeat"}[state]
        state_url = f"{PLACEHOLDER_BASE}/160x160/{RARITY_COLORS_HEX.get(char_data['rarity'], '888888')}/FFFFFF?text={char_data['name'][:8]}+{state_label}"[:210]
        if not dry_run:
            download_image(state_url, gif_path, retries=2)
        else:
            print(f"  [DRY-RUN] Baixaria GIF: {state_url} → {gif_path}")


def download_boss_assets(boss_id, boss_data, dry_run=False):
    """Baixa imagem e GIFs de um boss."""
    name = boss_data["name"].lower()
    diff = boss_data["difficulty"]
    print(f"\n👹 Boss {boss_data['name']} ({diff}) [{boss_data['anime']}]")

    # Imagem estática do boss
    img_path = IMAGES_DIR / f"boss_{name}.png"
    color = {"easy": "3DDC84", "normal": "4F9DDE", "hard": "B388FF", "nightmare": "FF3B3B"}.get(diff, "FF5E5B")
    url = f"{PLACEHOLDER_BASE}/160x160/{color}/FFFFFF?text={boss_data['name']}"[:180]
    if not dry_run:
        download_image(url, img_path)
    else:
        print(f"  [DRY-RUN] Baixaria: {url} → {img_path}")

    # GIFs de animação do boss
    for state in ["attack", "defend"]:
        gif_path = GIFS_DIR / f"boss_{name}_{state}.gif"
        state_url = f"{PLACEHOLDER_BASE}/160x160/{color}/FFFFFF?text=Boss+{state}"[:180]
        if not dry_run:
            download_image(state_url, gif_path, retries=2)
        else:
            print(f"  [DRY-RUN] Baixaria GIF: {state_url} → {gif_path}")


def download_sound_effects(dry_run=False):
    """Baixa sons de combate (placeholders)."""
    sounds = ["attack.wav", "defend.wav", "skill.wav", "hit.wav", "victory.wav", "defeat.wav"]
    for s in sounds:
        path = SOUNDS_DIR / s
        if not dry_run:
            # Só cria placeholder simples (sem download real, evita dependência)
            if not path.exists():
                # Cria um WAV silencioso mínimo (44 bytes header)
                header = bytes.fromhex("52494646240a000057415645666d7420100000000100010044ac0000885801000200100064617461")
                # Completa o header corretamente
                header = bytearray(44)
                header[0:4] = b"RIFF"
                header[4:8] = (len(header) + 0).to_bytes(4, "little")
                header[8:12] = b"WAVE"
                header[12:16] = b"fmt "
                header[16:20] = (16).to_bytes(4, "little")
                header[20:22] = (1).to_bytes(2, "little")  # PCM
                header[22:24] = (1).to_bytes(2, "little")  # mono
                header[24:28] = (22050).to_bytes(4, "little")  # sample rate
                header[28:32] = (22050).to_bytes(4, "little")  # byte rate
                header[32:34] = (1).to_bytes(2, "little")  # block align
                header[34:36] = (8).to_bytes(2, "little")  # bits per sample
                header[36:40] = b"data"
                header[40:44] = (0).to_bytes(4, "little")  # data size
                path.parent.mkdir(parents=True, exist_ok=True)
                path.write_bytes(bytes(header))
                print(f"  🔊 Som placeholder: {path.relative_to(BASE_DIR)}")
        else:
            print(f"  [DRY-RUN] Criaria som: {path.relative_to(BASE_DIR)}")


def main():
    print("=" * 55)
    print("  🌍 DOWNLOADER DE ASSETS DO GACHAGAME")
    print("=" * 55)

    args = sys.argv[1:]
    dry_run = "--dry-run" in args
    char_filter = None
    for i, a in enumerate(args):
        if a == "--character" and i + 1 < len(args):
            char_filter = int(args[i + 1]) if args[i + 1].isdigit() else None
            # ou nome
            for cid, cdata in CHARACTERS.items():
                if cdata["name"].lower() == args[i + 1].lower():
                    char_filter = cid
                    break

    conn = get_db()
    if conn:
        ensure_columns(conn)

    # Download personagens
    for cid, cdata in CHARACTERS.items():
        if char_filter and cid != char_filter:
            continue
        download_character_assets(cid, cdata, dry_run)
        if conn:
            update_character_urls(conn, cid, cdata)

    # Download bosses
    for bid, bdata in BOSSES.items():
        download_boss_assets(bid, bdata, dry_run)
        if conn:
            update_boss_urls(conn, bid, bdata)

    # Sons
    download_sound_effects(dry_run)

    if conn:
        conn.close()

    print("\n" + "=" * 55)
    print("  ✅ DOWNLOAD CONCLUÍDO!")
    print(f"  Imagens:  {IMAGES_DIR}")
    print(f"  GIFs:     {GIFS_DIR}")
    print(f"  Sons:     {SOUNDS_DIR}")
    print("=" * 55)


if __name__ == "__main__":
    main()

File: python/image_tools/download_assets.js
#!/usr/bin/env node
/**
 * 🌍 Baixa Imagens Reais dos Personagens (AniList) + Gera GIFs Locais
 *
 * Baixa imagens reais de personagens de anime da API pública do AniList,
 * gera GIFs de combate animados localmente (via Pillow/Python), e
 * atualiza o banco SQLite com as URLs locais.
 *
 * Uso:
 *   node python/image_tools/download_assets.js
 *   node python/image_tools/download_assets.js --dry-run
 */

const fs = require('fs');
const path = require('path');
const https = require('https');
const http = require('http');
const { execSync } = require('child_process');

const BASE_DIR = path.resolve(__dirname, '..', '..');
const DB_PATH = path.join(BASE_DIR, 'data', 'gacha-game.db');
const IMAGES_DIR = path.join(BASE_DIR, 'assets', 'images');
const GIFS_DIR = path.join(BASE_DIR, 'assets', 'gifs');
const SOUNDS_DIR = path.join(BASE_DIR, 'assets', 'sounds');
const TF_DB = path.join(BASE_DIR, 'data', 'gacha-game.db');

const ANILIST_API = 'https://graphql.anilist.co';
const WIKIPEDIA_API = 'https://en.wikipedia.org/w/api.php';
const COMMONS_API = 'https://commons.wikimedia.org/w/api.php';

// ====== Mapeamento de personajes → IDs del AniList ======
// ID del banco local → { file, anilist_id }
const CHARACTERS = [
  { db_id: 1, file: 'frieren', anilist_id: 176754 },
  { db_id: 2, file: 'gilgamesh', anilist_id: 2514 },
  { db_id: 3, file: 'madoka', anilist_id: 37832 },
  { db_id: 4, file: 'ichigo', anilist_id: 5 },
  { db_id: 5, file: 'johan', anilist_id: 719 },
  { db_id: 6, file: 'jotaro', anilist_id: 4003 },
  { db_id: 7, file: 'dio', anilist_id: 4004 },
  { db_id: 8, file: 'giorno', anilist_id: 10529 },
  { db_id: 9, file: 'rukia', anilist_id: 6 },
  { db_id: 10, file: 'aizen', anilist_id: 1086 },
  { db_id: 11, file: 'fern', anilist_id: 183965 },
  { db_id: 12, file: 'stark', anilist_id: 184313 },
  { db_id: 13, file: 'joseph', anilist_id: 6356 },
  { db_id: 14, file: 'kars', anilist_id: 21966 },
  { db_id: 15, file: 'uryu', anilist_id: 564 },
  { db_id: 16, file: 'orihime', anilist_id: 7 },
  { db_id: 17, file: 'eisen', anilist_id: 184312 },
  { db_id: 18, file: 'caesar', anilist_id: 21959 },
  { db_id: 19, file: 'lisa', anilist_id: 21960 },
  { db_id: 20, file: 'renji', anilist_id: 906 },
  { db_id: 21, file: 'sein', anilist_id: 205177 },
  { db_id: 22, file: 'speedwagon', anilist_id: 21938 },
  { db_id: 23, file: 'karin', anilist_id: 4018 },
  { db_id: 24, file: 'kraft', anilist_id: 219732 },
  { db_id: 25, file: 'erina', anilist_id: 21937 },
  { db_id: 26, file: 'yuzu', anilist_id: 1096 },
  { db_id: 27, file: 'heiter', anilist_id: 184310 },
  { db_id: 28, file: 'poco', anilist_id: 89479 },
];

// ====== Mapeamiento de bosses → IDs del AniList ======
const BOSSES = [
  { db_id: 1, file: 'boss_aizen', anilist_id: 1086 },
  { db_id: 2, file: 'boss_ulquiorra', anilist_id: 1081 },
  { db_id: 3, file: 'boss_grimmjow', anilist_id: 1080 },
  { db_id: 4, file: 'boss_frieren_dragon', anilist_id: 184313 }, // Dragão Ancião - usa Stark como referencia visual
  { db_id: 5, file: 'boss_qual', anilist_id: 315278 },
  { db_id: 6, file: 'boss_wolf', anilist_id: 176754 }, // Lobo Demoníaco - usa Frieren
  { db_id: 7, file: 'boss_kars', anilist_id: 21966 },
  { db_id: 8, file: 'boss_dio', anilist_id: 4004 },
  { db_id: 9, file: 'boss_acdc', anilist_id: 21964 },
];

const RARITY_COLORS = {
  common: '95A5A6', rare: '74B9FF', epic: '4F9DDE',
  legendary: 'B388FF', mythic: 'FFC93C', secret: 'FF3B3B',
};

const STATES = ['idle', 'attack', 'defend', 'skill', 'hit', 'victory', 'defeat'];

function ensureDir(dir) {
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
}

function download(url, filepath) {
  return new Promise((resolve) => {
    if (fs.existsSync(filepath) && fs.statSync(filepath).size > 1000) {
      console.log('  📥 (skip) ' + path.relative(BASE_DIR, filepath));
      return resolve(filepath);
    }
    const lib = url.startsWith('https') ? https : http;
    const req = lib.get(url, { headers: { 'User-Agent': 'GachaGame-AssetDownloader/1.0' } }, (res) => {
      const chunks = [];
      res.on('data', (c) => chunks.push(c));
      res.on('end', () => {
        const data = Buffer.concat(chunks);
        if (data.length > 1000) {
          ensureDir(path.dirname(filepath));
          fs.writeFileSync(filepath, data);
          console.log('  📥 ' + path.relative(BASE_DIR, filepath) + ' (' + data.length + ' bytes)');
          resolve(filepath);
        } else {
          console.log('  ❌ Falhou: ' + url);
          resolve(null);
        }
      });
    });
    req.on('error', () => resolve(null));
    req.setTimeout(15000, () => { req.destroy(); resolve(null); });
  });
}

async function anilistGetById(id) {
  const query = `
    query {
      Character(id: ${id}) {
        id
        name { full }
        image { large }
      }
    }
  `;
  try {
    const r = await fetch(ANILIST_API, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ query })
    });
    const d = await r.json();
    if (d.data && d.data.Character) {
      return d.data.Character;
    }
  } catch (e) {}
  return null;
}

async function anilistSearch(name) {
  const query = `
    query ($search: String) {
      Character(search: $search) {
        id
        name { full }
        image { large }
      }
    }
  `;
  try {
    const r = await fetch(ANILIST_API, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ query, variables: { search: name } })
    });
    const d = await r.json();
    if (d.data && d.data.Character) return d.data.Character;
  } catch (e) {}
  return null;
}

async function wikipediaImage(name) {
  const params = new URLSearchParams({
    action: 'query',
    titles: name.replace(/ /g, '_'),
    prop: 'pageimages',
    format: 'json',
    pithumbsize: '440',
    redirects: '1',
  });
  try {
    const response = await fetch(`${WIKIPEDIA_API}?${params}`, {
      headers: { 'User-Agent': 'GachaGameAssetDownloader/1.0 (local development)' },
    });
    const data = await response.json();
    const page = Object.values(data.query?.pages || {})[0];
    return page?.thumbnail?.source || null;
  } catch (e) {
    return null;
  }
}

async function commonsImage(name) {
  const params = new URLSearchParams({
    action: 'query',
    generator: 'search',
    gsrsearch: `${name} character -cosplay -costume -fanart`,
    gsrnamespace: '6',
    gsrlimit: '10',
    prop: 'imageinfo',
    iiprop: 'url',
    iiurlwidth: '440',
    format: 'json',
  });
  try {
    const response = await fetch(`${COMMONS_API}?${params}`, {
      headers: { 'User-Agent': 'GachaGameAssetDownloader/1.0 (local development)' },
    });
    const data = await response.json();
    const pages = Object.values(data.query?.pages || {});
    const page = pages.find((candidate) => !/cosplay|costume|fanart/i.test(candidate.title));
    return page?.imageinfo?.[0]?.thumburl || page?.imageinfo?.[0]?.url || null;
  } catch (e) {
    return null;
  }
}

async function generateGifForState(inputFile, outputFile, state) {
  return new Promise((resolve) => {
    try {
      const script = path.join(BASE_DIR, 'python', 'image_tools', 'combat_gif_generator.py');
      execSync(`python "${script}" "${inputFile}" "${outputFile}" ${state}`, { stdio: 'pipe' });
      resolve(outputFile);
    } catch (e) {
      console.log('  ❌ GIF falhou: ' + state + ' → ' + e.message);
      resolve(null);
    }
  });
}

async function main() {
  const args = process.argv.slice(2);
  const dry = args.includes('--dry-run');
  console.log('=' .repeat(60));
  console.log('  🌍 DOWNLOADER DE ASSETS REAIS (AniList + Pillow GIFs)');
  console.log('=' .repeat(60));

  // Conecta ao DB
  let db;
  try {
    db = require('better-sqlite3')(DB_PATH);
    // Garante colunas
    for (const table of ['characters', 'bosses']) {
      const cols = db.prepare(`PRAGMA table_info(${table})`).all().map(r => r.name);
      const needed = table === 'characters'
        ? ['image_url', 'image_idle_url', 'gif_attack_url', 'gif_defend_url', 'gif_skill_url', 'gif_hit_url', 'gif_victory_url', 'gif_defeat_url']
        : ['image_url', 'gif_attack_url', 'gif_defend_url', 'gif_skill_url'];
      for (const c of needed) {
        if (!cols.includes(c)) {
          db.prepare(`ALTER TABLE ${table} ADD COLUMN ${c} TEXT`).run();
          console.log(`  ➕ Coluna ${c} em ${table}`);
        }
      }
    }
  } catch (e) {
    console.log('⚠️  Não conectou ao DB:', e.message);
  }

  const AP = '/assets/';

  // ====== Personagens ======
  console.log('\n' + '─'.repeat(60));
  console.log('  🧙 PERSONAGENS');
  console.log('─'.repeat(60));

  for (const c of CHARACTERS) {
    console.log('\n🔧 ' + c.file + ' (db_id=' + (c.db_id || 'by name') + ')');

    if (!dry) {
      // Busca URL real do AniList
      const anilistCharacter = c.anilist_id
        ? await anilistGetById(c.anilist_id)
        : await anilistSearch(c.search);
      const imageUrl = anilistCharacter?.image?.large
        || await wikipediaImage(c.search || c.file)
        || await commonsImage(c.search || c.file);
      if (!imageUrl) {
        console.log('  ❌ Imagem não encontrada em AniList, Wikipédia ou Commons: ' + c.file);
        continue;
      }
      console.log('  🎯 Fonte da imagem: ' + (anilistCharacter ? 'AniList' : 'Wikipédia/Commons'));
      
      // Baixa imagem estática
      const imgPath = path.join(IMAGES_DIR, c.file + '.png');
      await download(imageUrl, imgPath);
      if (!fs.existsSync(imgPath) || fs.statSync(imgPath).size < 1000) {
        console.log('  ⚠️ Falha ao baixar imagem, usando placeholder');
        continue;
      }

      // Gera GIFs locais
      for (const state of STATES) {
        const gifPath = path.join(GIFS_DIR, c.file + '_' + state + '.gif');
        if (!fs.existsSync(gifPath) || fs.statSync(gifPath).size < 1000) {
          await generateGifForState(imgPath, gifPath, state);
        } else {
          console.log('  📥 (skip) ' + path.relative(BASE_DIR, gifPath));
        }
      }

      // Atualiza banco
      if (db) {
        const dbCharacter = c.db_id
          ? { id: c.db_id }
          : db.prepare('SELECT id FROM characters WHERE name = ?').get(c.name);
        if (!dbCharacter) {
          console.log('  ❌ Personagem não encontrado no banco: ' + c.name);
          continue;
        }
        const updates = {
          image_url: AP + 'images/' + c.file + '.png',
          image_idle_url: AP + 'images/' + c.file + '.png',
          gif_attack_url: AP + 'gifs/' + c.file + '_attack.gif',
          gif_defend_url: AP + 'gifs/' + c.file + '_defend.gif',
          gif_skill_url: AP + 'gifs/' + c.file + '_skill.gif',
          gif_hit_url: AP + 'gifs/' + c.file + '_hit.gif',
          gif_victory_url: AP + 'gifs/' + c.file + '_victory.gif',
          gif_defeat_url: AP + 'gifs/' + c.file + '_defeat.gif',
        };
        const set = Object.keys(updates).map(k => k + ' = ?').join(', ');
        db.prepare('UPDATE characters SET ' + set + ' WHERE id = ?').run(...Object.values(updates), dbCharacter.id);
        console.log('  📝 Banco atualizado: ' + c.file);
      }
    } else {
      console.log('  [DRY-RUN] Baixaria imagem + GIFs para: ' + c.file);
    }

    // Rate limit AniList
    await new Promise(r => setTimeout(r, 700));
  }

  // ====== Bosses ======
  console.log('\n' + '─'.repeat(60));
  console.log('  👹 BOSSES');
  console.log('─'.repeat(60));

  for (const b of BOSSES) {
    console.log('\n👹 ' + b.file + ' (db_id=' + b.db_id + ')');

    if (!dry) {
      const boss = await anilistGetById(b.anilist_id);
      if (!boss) {
        console.log('  ❌ Boss não encontrado no AniList: ' + b.file);
        continue;
      }
      console.log('  🎯 ' + boss.name.full);

      const imgPath = path.join(IMAGES_DIR, b.file + '.png');
      await download(boss.image.large, imgPath);
      if (!fs.existsSync(imgPath) || fs.statSync(imgPath).size < 1000) {
        console.log('  ⚠️ Falha ao baixar imagem do boss');
        continue;
      }

      // GIFs de boss (attack, defend, skill)
      for (const state of ['attack', 'defend', 'skill']) {
        const gifPath = path.join(GIFS_DIR, b.file + '_' + state + '.gif');
        if (!fs.existsSync(gifPath) || fs.statSync(gifPath).size < 1000) {
          await generateGifForState(imgPath, gifPath, state);
        } else {
          console.log('  📥 (skip) ' + path.relative(BASE_DIR, gifPath));
        }
      }

      // Atualiza banco
      if (db) {
        db.prepare('UPDATE bosses SET image_url = ?, gif_attack_url = ?, gif_defend_url = ?, gif_skill_url = ? WHERE id = ?').run(
          AP + 'images/' + b.file + '.png',
          AP + 'gifs/' + b.file + '_attack.gif',
          AP + 'gifs/' + b.file + '_defend.gif',
          AP + 'gifs/' + b.file + '_skill.gif',
          b.db_id
        );
        console.log('  📝 Banco atualizado: ' + b.file);
      }
    } else {
      console.log('  [DRY-RUN] Baixaria imagem + GIFs para: ' + b.file);
    }

    await new Promise(r => setTimeout(r, 700));
  }

  // ====== Sons ======
  if (!dry) {
    ensureDir(SOUNDS_DIR);
    for (const s of ['attack', 'defend', 'skill', 'hit', 'victory', 'defeat']) {
      const p = path.join(SOUNDS_DIR, s + '.wav');
      if (!fs.existsSync(p)) {
        const h = Buffer.alloc(44);
        h.write('RIFF', 0); h.write('WAVE', 8); h.write('fmt ', 12);
        h.writeUInt32LE(16, 16); h.writeUInt16LE(1, 20); h.writeUInt16LE(1, 22);
        h.writeUInt32LE(22050, 24); h.writeUInt32LE(22050, 28); h.writeUInt16LE(1, 32);
        h.writeUInt16LE(8, 34); h.write('data', 36); h.writeUInt32LE(0, 40);
        fs.writeFileSync(p, h);
        console.log('  🔊 Som placeholder: ' + path.relative(BASE_DIR, p));
      }
    }
  }

  if (db) db.close();
  console.log('\n' + '='.repeat(60));
  console.log('  ✅ DOWNLOAD CONCLUÍDO!');
  console.log('  Imagens: ' + IMAGES_DIR);
  console.log('  GIFs:    ' + GIFS_DIR);
  console.log('  Sons:    ' + SOUNDS_DIR);
  console.log('='.repeat(60));
}

main().catch((e) => { console.error(e); process.exit(1); });